content/apa-itu-domain-driven-design.md

```diff
@@ -0,0 +1,79 @@
+sebuah domain yang berdiri sendiri, walaupun nanti sebuah domain dapat mempunyai ketergantungan (bahasa inggris: dependent)
+kepada domain lain. Misalnya untuk mengeksekusi suatu fungsi dari domain absensi, harus mengambil beberapa data dari data
+guru & murid. Namun, masing-masing domain juga memiliki aturan dan batasan untuk hal-hal yang dapat dilakukan.
+
+Dalam domain-driven design, seorang developer nggak bisa dilepas sendiri untuk develop sebuah domain tanpa mengetahui
+business requirement tanpa bicara dengan _domain expert_. Wah, istilah baru lagi. Domain expert adalah seseorang yang mengerti
+dengan ranah bisnis dari suatu domain.
+
+Setelah ngobrol-ngobrol dengan domain expert, dalam menciptakan suatu model domain, biasanya diperlukan sebuah konteks.
+Konteks dapat berupa event, statement, ide, atau apapun itu yang mempunyai ikatan dengan domain. Biasanya, seorang developer
+akan kepikiran untuk menciptakan model untuk domainnya. Dalam suatu domain, sangat umum juga apabila terdapat lebih dari
+satu model. Dalam setiap model pasti memiliki aturan dan batasannya sendiri, begitupula dengan implementasinya.
+
+Menghadapi banyak model secara bersamaan akan menjadi suatu masalah. Oleh karena itu, harus terdapat _ubiquitous language_,
+yang bearti penamaan (bahasa inggris: naming convention) atas segala sesuatu yang ada di dalam konteks tersebut harus
+manusiawi dan lebih dekat dengan bisnis. Nggak boleh disingkat-singkat atau bahkan misleading.
+
+Sampai titik ini kamu seharusnya sadar bahwa domain-driven design itu tidak berfokus kepada developers saja, namun kepada
+semua orang hingga tim bisnis.
+
+Lalu bagaimana secara arsitektur?
+
+Dalam sebuah aplikasi yang kompleks, biasanya berbagai hal seperti UI, database, sampai business logic dipisahkan kedalam
+beberapa layer, yang biasanya membantu developer untuk me-manage kompleksitas kodenya. Secara dogmatis, domain-driven design
+menyarankan kita untuk memisahkan ekspresi antara model dengan business logic. Buat sebuah desain arsitektur yang mana
+masing-masing layer hanya mempunyai ketergantungan terhadap layer dibawahnya. Sehingga kode tersebut menjadi loosely-coupled,
+yang mana berarti terpisah antara UI, aplikasi, dan kode infrastruktur. Ada beberapa istilah dalam _layered architecture_:
+
+1. _Entity_, yang merujuk kepada sebuah object yang memiliki identitas, punya behavior, aturan, dan side effect. Dimana
+   identitas dari suatu entity ini tidak akan pernah berubah, namun property lain di dalamnya bisa berubah
+   (atau ter-mutate).
+2. _Value object_ merupakan sebuah object yang tidak memiliki indentitas, lebih sederhana lagi dapat didefinisikan sebagai
+   atribut dan logika dari sebuah elemen dalam model. Value object ini dapat berubah, namun seharusnya diperlakukan sebagai
+   suatu object yang immutable (atau isinya tidak bisa diubah). Apabila berubah, suatu value object baru akan diciptakan
+   dan di-assign kepada layer diatasnya yang biasanya merupakan sebuah entity.
```

Comment:
ada padanan kata yang lebih familiar dari _assign_, seperti diberikan , ini ga literal translate dari inggris tapi lebih menegaskan ke arah semantik.

Reply:
kurang cocok sih antara assign dan diberikan. kalo diberikan kan secara semantik lebih cuma "give" something to someone, sedangkan kalo assign ngga cuma ngasih, tapi minta dikerjakan oleh mereka juga

Reply:
kan di kasih buat dikerjain hehe